File: test/test_zod2ts/class-expected/layered_generics.infra.ts
// This is
// a multiline
// header.

import * as GENERICS_APP from "./layered_generics.app";

export class HttpSuccessfulResponse<T> {
    success: true;
    data: T;

    constructor(data: HttpSuccessfulResponse<T>) {
        this.success = data.success;
        this.data = data.data;
    }
}

export class HttpUnsuccessfulResponse {
    success: false;
    message: string;
    details?: Record<string, any>;

    constructor(data: HttpUnsuccessfulResponse) {
        this.success = data.success;
        this.message = data.message;
        this.details = data.details;
    }
}

export class HttpErrorResponse {
    message: string;

    constructor(data: HttpErrorResponse) {
        this.message = data.message;
    }
}

export class SomeDtoResult {
    id: string;
    name: string;
    age: number;

    constructor(data: SomeDtoResult) {
        this.id = data.id;
        this.name = data.name;
        this.age = data.age;
    }
}

export class InternalObjectWithGeneric extends HttpSuccessfulResponse<SomeDtoResult> {}

export class ObjectWithGeneric {
    internal: InternalObjectWithGeneric;
    item: HttpSuccessfulResponse<SomeDtoResult>;
    userItem: GENERICS_APP.GenericUserEntity<SomeDtoResult>;
    otherUserItem: GENERICS_APP.AdminUserEntity;

    constructor(data: ObjectWithGeneric) {
        this.internal = data.internal;
        this.item = data.item;
        this.userItem = data.userItem;
        this.otherUserItem = data.otherUserItem;
    }
}

export class OtherDtoResult {
    code: string;
    description: string;

    constructor(data: OtherDtoResult) {
        this.code = data.code;
        this.description = data.description;
    }
}

export class DataRetrieve extends HttpSuccessfulResponse<SomeDtoResult> {}

export type DiscriminantDataRetrieve =
    | HttpSuccessfulResponse<SomeDtoResult>
    | HttpUnsuccessfulResponse;

// Built from intersection of HttpSuccessfulResponse and GenericUserEntity
export class IntersectedDataRetrieve {
    success: true;
    data: SomeDtoResult;
    id: string;
    name: string;
    email: string;
    age?: number;
    metadata: OtherDtoResult;

    constructor(data: IntersectedDataRetrieve) {
        this.success = data.success;
        this.data = data.data;
        this.id = data.id;
        this.name = data.name;
        this.email = data.email;
        this.age = data.age;
        this.metadata = data.metadata;
    }
}

export class UserRetrieve extends HttpSuccessfulResponse<GENERICS_APP.NormalUserEntity> {}

File: test/test_zod2go/struct-expected/layered_generics.infra.go
// This is
// a multiline
// header.

package models

import (
	"encoding/json"
	"fmt"
	GENERICS_APP "./layered_generics.app"
)

type HttpSuccessfulResponse[T any] struct {
    Success bool `json:"success"`
    Data T `json:"data"`
}

type HttpUnsuccessfulResponse struct {
    Success bool `json:"success"`
    Message string `json:"message"`
    Details *map[string]any `json:"details,omitempty"`
}

type HttpErrorResponse struct {
    Message string `json:"message"`
}

type SomeDtoResult struct {
    Id string `json:"id"`
    Name string `json:"name"`
    Age int64 `json:"age"`
}

type InternalObjectWithGeneric struct {
    HttpSuccessfulResponse[SomeDtoResult]
}

type ObjectWithGeneric struct {
    Internal InternalObjectWithGeneric `json:"internal"`
    Item HttpSuccessfulResponse[SomeDtoResult] `json:"item"`
    UserItem GENERICS_APP.GenericUserEntity[SomeDtoResult] `json:"userItem"`
    OtherUserItem GENERICS_APP.AdminUserEntity `json:"otherUserItem"`
}

type OtherDtoResult struct {
    Code string `json:"code"`
    Description string `json:"description"`
}

type DataRetrieve struct {
    HttpSuccessfulResponse[SomeDtoResult]
}

// DiscriminantDataRetrieve is a discriminated union on "success".
// Possible types: HttpSuccessfulResponse[SomeDtoResult], HttpUnsuccessfulResponse
type DiscriminantDataRetrieve interface {
    isDiscriminantDataRetrieve()
}

func (t HttpSuccessfulResponse[T]) isDiscriminantDataRetrieve() {}

func (t HttpUnsuccessfulResponse) isDiscriminantDataRetrieve() {}

// UnmarshalDiscriminantDataRetrieve deserializes JSON into the correct DiscriminantDataRetrieve concrete type
// by probing the "success" discriminant field.
func UnmarshalDiscriminantDataRetrieve(data []byte) (DiscriminantDataRetrieve, error) {
    var probe struct {
        Success json.RawMessage `json:"success"`
    }
    if err := json.Unmarshal(data, &probe); err != nil {
        return nil, err
    }
    switch string(probe.Success) {
    case "true":
        var v HttpSuccessfulResponse[SomeDtoResult]
        if err := json.Unmarshal(data, &v); err != nil {
            return nil, err
        }
        return v, nil
    case "false":
        var v HttpUnsuccessfulResponse
        if err := json.Unmarshal(data, &v); err != nil {
            return nil, err
        }
        return v, nil
    }
    return nil, fmt.Errorf("failed to deserialize DiscriminantDataRetrieve: unknown discriminator %s", string(probe.Success))
}

type IntersectedDataRetrieve struct {
    Success bool `json:"success"`
    Data SomeDtoResult `json:"data"`
    Id string `json:"id"`
    Name string `json:"name"`
    Email string `json:"email"`
    Age *int64 `json:"age,omitempty"`
    Metadata OtherDtoResult `json:"metadata"`
}

type UserRetrieve struct {
    HttpSuccessfulResponse[GENERICS_APP.NormalUserEntity]
}
